File: sdk/cpp/src/api/catalog.cpp
// Copyright 2026 Hypermesh Foundation. All rights reserved.
// Licensed under the Business Source License 1.1.

#include <hypermesh/api/catalog.hpp>

namespace hypermesh {

CatalogPackageList CatalogApi::browse(const std::string& query,
                                       uint32_t page) const {
    std::string path = "/api/v1/catalog/browse";
    std::string sep = "?";
    if (!query.empty()) {
        path += sep + "query=" + query;
        sep = "&";
    }
    if (page > 0) {
        path += sep + "page=" + std::to_string(page);
    }
    auto j = http_.get(path);
    return j.get<CatalogPackageList>();
}

CatalogSearchResults CatalogApi::search(const std::string& query) const {
    auto j = http_.get("/api/v1/catalog/search?query=" + query);
    return j.get<CatalogSearchResults>();
}

CatalogPackageInfo CatalogApi::package_info(const std::string& name) const {
    auto j = http_.get("/api/v1/catalog/package/" + name);
    return j.get<CatalogPackageInfo>();
}

CatalogRegistryStats CatalogApi::registry_stats() const {
    auto j = http_.get("/api/v1/catalog/registry/stats");
    return j.get<CatalogRegistryStats>();
}

} // namespace hypermesh

File: sdk/typescript/src/api/catalog.ts
import type { HttpClient } from "../client.js";
import type {
  CatalogPackageInfo,
  CatalogPackageList,
  CatalogRegistryStats,
  CatalogSearchResults,
} from "../types.js";

export class CatalogApi {
  constructor(private readonly http: HttpClient) {}

  async browse(query?: string, page?: number): Promise<CatalogPackageList> {
    const params: string[] = [];
    if (query !== undefined) params.push(`query=${encodeURIComponent(query)}`);
    if (page !== undefined) params.push(`page=${page}`);
    const qs = params.length > 0 ? `?${params.join("&")}` : "";
    return this.http.get<CatalogPackageList>(
      `/api/v1/catalog/browse${qs}`,
    );
  }

  async search(query: string): Promise<CatalogSearchResults> {
    return this.http.get<CatalogSearchResults>(
      `/api/v1/catalog/search?query=${encodeURIComponent(query)}`,
    );
  }

  async packageInfo(name: string): Promise<CatalogPackageInfo> {
    return this.http.get<CatalogPackageInfo>(
      `/api/v1/catalog/package/${encodeURIComponent(name)}`,
    );
  }

  async registryStats(): Promise<CatalogRegistryStats> {
    return this.http.get<CatalogRegistryStats>(
      "/api/v1/catalog/registry/stats",
    );
  }
}
